### DISCUSSION TITLE
Support nativeaot?

### DISCUSSION THREAD

**Discussion by leveljiujiu** [friflo/Friflo.Json.Fliox#39] [Q&A] [answered]
This is a great project, my only question is whether it supports native aot ?

**Comment by friflo** [comment]
Hi,
I never tried. So I did it right now.
I followed the steps at:
[Publish Native AOT using the CLI](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/)

Result with v1.24.0. It fails with
```
Unhandled Exception: System.TypeInitializationException: A type initializer threw an exception. To determine which type, inspect the InnerException's StackTrace property.
 ---> System.PlatformNotSupportedException: Operation is not supported on this platform.
   at System.Reflection.Runtime.Assemblies.RuntimeAssemblyInfo.GetReferencedAssemblies() + 0x24
   at Friflo.Engine.ECS.AssemblyLoader.CheckAssembly(Assembly) + 0x4e
   at Friflo.Engine.ECS.AssemblyLoader.GetEngineDependants() + 0xb1
   at Friflo.Engine.ECS.SchemaUtils.RegisterSchemaTypes(TypeStore) + 0x3e
   at Friflo.Engine.ECS.EntityStoreBase.Static..cctor() + 0x61
   at System.Runtime.CompilerServices.ClassConstructorRunner.EnsureClassConstructorRun(StaticClassConstructionContext*) + 0xb9
   --- End of inner exception stack trace ---
   at System.Runtime.CompilerServices.ClassConstructorRunner.EnsureClassConstructorRun(StaticClassConstructionContext*) + 0x14a
   at System.Runtime.CompilerServices.ClassConstructorRunner.CheckStaticClassConstructionReturnGCStaticBase(StaticClassConstructionContext*, Object) + 0xd
   at Friflo.Engine.ECS.EntityStoreBase..ctor() + 0x13d
   at Friflo.Engine.ECS.EntityStore..ctor(PidType) + 0x4c
   at GameTest.ParallelQuery.Query_ForEach(String[]) + 0x2e4
```

I had NativeAOT in mind when developing this project.
So I avoided using [Limitations of Native AOT deployment](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/#limitations-of-native-aot-deployment)

The only exception I see right now is using `RuntimeAssemblyInfo`.
I will check what needs to be done to get it compatible with NativeAOT.

**Comment by friflo** [comment]
1. Fiddled around to get something working.
- Added RegisterComponent(), RegisterTag() and RegisterScript() methods.
- Call the Register*() methods in Main().
- Disabled JSON Serialization.

With these changes Native AOT is running.

2. Issues that need to be workaround in Native AOT.
- Assembly.GetTypes() cannot be used in Native AOT. It returns on 6 unrelated types.
  Therefore the Register() methods are added.
  A Roslyn Analyzer could be implemented to do this registration automatically.
- JSON Serialization is currently reflection based.
  So all types that could be serialized must not be trimmed.
  It would require adding a Roslyn Analyzer and a source generator. This is a big effort.

All in all only 1. is an option that can be implemented with reasonable effort.
But I do not really like it because of it requirements and limitations.

@leveljiujiu What is your use case?

**Reply by leveljiujiu** [reply]
Personally, I think 1 is a good choice. I have no need for json serialization. We can keep the pattern and leave the 1 as an extension.

**Reply by leveljiujiu** [reply]
Because my future project needs to release iOS platform, so I am more concerned about nativeaot.

**Reply by friflo** [reply]
Okay, I will add the feature using approach 1.
I guess it is not urgent. Ping me again if I forget it.
See: https://github.com/friflo/Friflo.Json.Fliox/issues/40

**Reply by leveljiujiu** [reply]
Thanks, adding it to the roadmap is enough.

**Reply by friflo** [reply]
I started working on this topic. Its running already on my local system.
I plan to publish NativeAOT support tomorrow.

**Comment by friflo** [comment, marked as answer]
hey @leveljiujiu, @srdw
Just published a new package with Native AOT support.

How to use is documented at
See [Wiki · Examples ~ General](https://github.com/friflo/Friflo.Json.Fliox/wiki/Examples-~-General#native-aot)
